File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        sol = []
        index = 0
        while index < len(nums) - 1:
            target = nums[index] * -1
            l, r = index + 1, len(nums)-1
            while l < r:
                if nums[l] + nums[r] == target:
                    sol.append([nums[l], nums[r], nums[index]])
                    l += 1
                    while l < r and nums[l-1] == nums[l]:
                        l += 1
                    r -= 1
                    while l < r and nums[r] == nums[r+1]: 
                        r -= 1
                elif nums[l] + nums[r] < target:
                    l += 1
                    while l < r and nums[l-1] == nums[l]:
                        l += 1
                else:
                    r -= 1
                    while l < r and nums[r] == nums[r+1]: 
                        r -= 1
            index += 1
            while index < len(nums) and nums[index-1] == nums[index]:
                index += 1
        return sol

File: Data Structures & Algorithms/partition-to-k-equal-sum-subsets/submission-0.java
class Solution {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        int totalSum = Arrays.stream(nums).sum();
        if(totalSum % k != 0){
            return false;
        }
        int targetSum = totalSum / k;
        Arrays.sort(nums);
        reverse(nums);

        return backtrack(nums, new int[k], 0, targetSum);
    }

    public boolean backtrack(int[] nums, int[] subsets, int index, int targetSum){
        if(index == nums.length){
            return true;
        }

        for(int k = 0; k < subsets.length; k++){
            if(subsets[k] + nums[index] <= targetSum){
                subsets[k] += nums[index];
                if(backtrack(nums, subsets, index + 1, targetSum)) return true;
                subsets[k] -= nums[index];
            }
            if(subsets[k] == 0){
                break;
            }
        }

        return false;
    }

    public void reverse(int[] nums){
        for (int i = 0, j = nums.length - 1; i < j; i++, j--) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }
}